world-chain-builder/crates/world/payload/src/builder.rs

```diff
@@ -644,7 +643,18 @@ where
             }
 
             let gas_used = match builder.execute_transaction(tx.clone()) {
-                Ok(res) => res,
+                Ok(res) => {
+                    if let Some(payloads) = pooled_tx.pbh_payload() {
+                        if spent_nullifier_hashes.is_empty() {
+                            block_gas_limit -= FIXED_GAS
+                        }
```

Comment:
An alternative approach is to optimistically reduce the `block_gas_limit` by `FIXED_GAS`. This assumes that the block will always have at least one PBH tx and in the case where it does not, the block will have `21000` less gas capacity.